### PR TITLE
k4e design description

### PR DESCRIPTION
**Pull request by jakub-dzon**
This PR documents k4e architecture and behavior design.

Signed-off-by: Jakub Dzon <jdzon@redhat.com>

### CODE REVIEW

File: docs/design.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+Object Storage is used to store files created by workloads on devices and uploaded using k4e built-in mechanism.
+
+Object Bucket Claims provided by OCS and Noobaa and accessible with like S3 buckets. 
+
+See [Data Upload](data-upload.md) for more details.
+
+#### CRDs
+
+Operator manages two kinds of CRs:
```
Comment (pkliczewski) [inline docs/design.md:25]: What do you think about linking to https://github.com/jakub-dzon/k4e-operator/tree/main/config/crd/bases both CRDs and https://github.com/jakub-dzon/k4e-operator/tree/main/config/samples?

Comment (jakub-dzon) [reply docs/design.md:25]: Done

**Review by pkliczewski — APPROVED** [review]

File: docs/workload_deployment.puml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+title workload deployment
+autonumber
+
+actor User
+
+User -> "EdgeDeployment CR": Create
+Operator -> "EdgeDeployment CR": Process
+Operator -> "EdgeDevice CR": add workload reference to status
+Operator -> Agent: Send pod yaml
```
Comment (masayag) [inline docs/workload_deployment.puml:12]: perhaps "respond with pod yaml" to emphasis there is no data being send from the operator to the agent ?

**Comment by masayag** [comment]
@jakub-dzon could you describe also (or add a link) to the steps for generating the image from the .puml files?

File: docs/workload_deployment.puml (hunk 2 above)
Comment (jakub-dzon) [reply docs/workload_deployment.puml:12]: Well, it is pulled by the agent - the pod yaml becomes part of the device configuration.

**Comment by jakub-dzon** [comment]
@masayag it's PlantUML tool, but I used IntelliJ Idea plugin for generating images:) 

CLI command is described here: https://plantuml.com/command-line.

File: docs/design.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+ 8. Agent starts processing requests (downloading configuration)
+ 9. Agent schedules workload data directories monitoring
+ 10. Agent schedules periodical heartbeat messages
+
+### Worklad deployment workflow
+
+![](workload_deployment.png)
+
+ 1. User creates `WorkloadDeployment` resource
+ 2. Operator process `WorkloadDeployment` resource and finds matching `EdgeDevices`
```
Comment (masayag) [inline docs/design.md:71]: could you add a paragraph for scheduling (using deviceID for pinning deployment to a specific device vs using the deviceSelector) ?

Comment (jakub-dzon) [reply docs/design.md:71]: There is PR #16 including that description.

File: docs/pairing.puml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+User -> Device: Boot k4e device ISO
+Device -> Agent: Start agent
+Agent -> Operator: Send pairing request
+Operator -> "EdgeDevice CR": Create
+Operator -> Agent: Pairing confirmation
+Operator -> OBC: Create Object Bucket Claim for the device data
+Operator -> "EdgeDevice CR": Update status with OBC info
+Agent -> Agent: Start processing requests
+Agent -> Clock: Start data directories monitoring
```
Comment (masayag) [inline docs/pairing.puml:17]: What does 'Clock' represents ?

Comment (jakub-dzon) [reply docs/pairing.puml:17]: Just an abstract clock (timer) that triggers some periodic actions.

**Review by masayag — APPROVED** [review]

**Pull request merged**